src/mod_commands.erl

```diff
@@ -453,15 +488,51 @@ run_subscription(Type, CallerJid, OtherJid) ->
     ejabberd_router:route(CallerJid, OtherJid, Acc2),
     ok.
 
-set_subscription(Caller, Other, <<"connect">>) ->
+
+set_subscription(Caller, Other, Action) ->
+    case check_jids(Caller, Other) of
+        ok ->
+            case Action of
+                A when A == <<"connect">>; A == <<"disconnect">> ->
+                    do_set_subscription(Caller, Other, Action);
+                _ ->
+                    {error, bad_request, <<"invalid action">>}
+            end;
+        E ->
+            E
+    end.
+
+do_set_subscription(Caller, Other, <<"connect">>) ->
     add_contact(Caller, Other),
     add_contact(Other, Caller),
     subscription(Caller, Other, <<"subscribe">>),
     subscription(Other, Caller, <<"subscribe">>),
     subscription(Other, Caller, <<"subscribed">>),
     subscription(Caller, Other, <<"subscribed">>),
     ok;
-set_subscription(Caller, Other, <<"disconnect">>) ->
+do_set_subscription(Caller, Other, <<"disconnect">>) ->
     delete_contact(Caller, Other),
     delete_contact(Other, Caller),
     ok.
+
+check_jids(CallerJid, OtherJid) ->
+    case {jid:from_binary(CallerJid), jid:from_binary(OtherJid)} of
+        {C, J} when C == error; J == error ->
+            {error, bad_request, <<"invalid jid">>};
+        _ ->
+            ok
+    end.
+
+parse_jid(Sender, Recipient) ->
+    case {parse_jid(Sender), parse_jid(Recipient)} of
+        {{error, Msg}, _} -> {error, type_error, Msg};
+        {_, {error, Msg}} -> {error, type_error, Msg};
+        {S, R} -> {ok, S, R}
+    end.
+
+parse_jid(Jid) ->
+    case jid:from_binary(Jid) of
+        error -> {error, io_lib:format("Invalid jid: ~p", [Jid])};
+        B -> B
+    end.
+
```

Comment:
I think that this function overloading is a bit of a bad style, same name but one parses a jid and the other parses sender and recipient. Calling the `/2` version `parse_jids`, in plural, would already be a better naming, if no other better idea is given.